modules/cudastereo/src/cuda/stereosgm_horizontal_path_aggregation.cu

```diff
@@ -0,0 +1,245 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include <cstdio>
+#include "stereosgm_horizontal_path_aggregation.hpp"
+#include "stereosgm_path_aggregation_common.hpp"
+#include "opencv2/core/cuda_stream_accessor.hpp"
+
+namespace cv { namespace cuda { namespace device
+{
+namespace stereosgm
+{
+namespace
+{
+static constexpr unsigned int DP_BLOCK_SIZE = 8u;
+static constexpr unsigned int DP_BLOCKS_PER_THREAD = 1u;
+
+static constexpr unsigned int WARPS_PER_BLOCK = 4u;
+static constexpr unsigned int BLOCK_SIZE = WARP_SIZE * WARPS_PER_BLOCK;
+
+
+template <int DIRECTION, unsigned int MAX_DISPARITY>
+__global__ void aggregate_horizontal_path_kernel(
+    PtrStep<int32_t> left,
+    PtrStep<int32_t> right,
+    PtrStep<uint8_t> dest,
+    int width,
+    int height,
+	unsigned int p1,
+	unsigned int p2)
+{
+	static const unsigned int SUBGROUP_SIZE = MAX_DISPARITY / DP_BLOCK_SIZE;
+	static const unsigned int SUBGROUPS_PER_WARP = WARP_SIZE / SUBGROUP_SIZE;
+	static const unsigned int PATHS_PER_WARP =
+		WARP_SIZE * DP_BLOCKS_PER_THREAD / SUBGROUP_SIZE;
+	static const unsigned int PATHS_PER_BLOCK =
+		BLOCK_SIZE * DP_BLOCKS_PER_THREAD / SUBGROUP_SIZE;
+
+	static_assert(DIRECTION == 1 || DIRECTION == -1, "");
+	if(width == 0 || height == 0){
+		return;
+	}
+
+	int32_t right_buffer[DP_BLOCKS_PER_THREAD][DP_BLOCK_SIZE];
+	DynamicProgramming<DP_BLOCK_SIZE, SUBGROUP_SIZE> dp[DP_BLOCKS_PER_THREAD];
+
+	const unsigned int warp_id  = threadIdx.x / WARP_SIZE;
+	const unsigned int group_id = threadIdx.x % WARP_SIZE / SUBGROUP_SIZE;
+	const unsigned int lane_id  = threadIdx.x % SUBGROUP_SIZE;
+	const unsigned int shfl_mask =
+		((1u << SUBGROUP_SIZE) - 1u) << (group_id * SUBGROUP_SIZE);
+
+	const unsigned int y0 =
+		PATHS_PER_BLOCK * blockIdx.x +
+		PATHS_PER_WARP  * warp_id +
+		group_id;
+	const unsigned int feature_step = SUBGROUPS_PER_WARP;
+	const unsigned int dest_step = SUBGROUPS_PER_WARP * MAX_DISPARITY * width;
+	const unsigned int dp_offset = lane_id * DP_BLOCK_SIZE;
+	left  = PtrStep<int32_t>( left.ptr(y0),  left.step);
+	right = PtrStep<int32_t>(right.ptr(y0), right.step);
+    dest  = PtrStep<uint8_t>(&dest(0, y0 * width * MAX_DISPARITY), dest.step);
+
+	if(y0 >= height){
+		return;
+	}
+
+	if(DIRECTION > 0){
+		for(unsigned int i = 0; i < DP_BLOCKS_PER_THREAD; ++i){
+			for(unsigned int j = 0; j < DP_BLOCK_SIZE; ++j){
+				right_buffer[i][j] = 0;
+			}
+		}
+	}else{
+		for(unsigned int i = 0; i < DP_BLOCKS_PER_THREAD; ++i){
+			for(unsigned int j = 0; j < DP_BLOCK_SIZE; ++j){
+				const int x = static_cast<int>(width - (j + dp_offset));
+				if(0 <= x && x < static_cast<int>(width)){
+					right_buffer[i][j] = __ldg(&right(i * feature_step, x));
```

Comment:
Do we have to avoid using `__ldg`?

Reply:
I searched about `__ldg` at [original repo](https://github.com/opencv/opencv_contrib), but I could not find it.

Reply:
Anyway, I think we have to use preprocessor which looks like `#if __CUDA_ARCH__ >= 350`.

Reply:
commit 130232d may resolve it.